### PR TITLE
Add regular expression operator

### PR DESCRIPTION
**Pull request by rfhb**
Hi, this is a proposal to add the operator `REGEXP` for use in SQL statements. It takes into account review comments in PR #266. 

The implementation here adds a single source file from [https://sqlite.org/src/raw?filename=ext/misc/regexp.c&ci=trunk](https://sqlite.org/src/raw?filename=ext/misc/regexp.c&ci=trunk), when `upgrade.R`is run. `Makevars` includes changes that are necessary to build it as a separate shared object; this tested successfully across platforms. The regular expression operator is loaded from the shared object by `RSQLite::initRegExp()`. The function documentation includes a very simple example; a couple of tests have been added in `test-regularExpression.R`. 

Thanks for reviewing and considering - I trust this is of general interest.

### CODE REVIEW

File: R/regularExpressions.R [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+         call. = FALSE)
+  }
+
+  lib_path <- getLoadedDLLs()[["RSQLite"]][["path"]]
+  # replace right-most name of library,
+  # should work across platforms
+  lib_path <- sub("(.*)RSQLite[.]", "\\1regexp.", lib_path)
+
+  # repeat loading would throw error
+  res <- try(
```
Comment (krlmlr) [inline R/regularExpressions.R]: Can we query if this extension is loaded?

File: R/regularExpressions.R [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+initRegExp <- function(db) {
+  if (!db@loadable.extensions) {
+    stop("Loadable extensions are not enabled for this db connection",
+         call. = FALSE)
+  }
+
+  lib_path <- getLoadedDLLs()[["RSQLite"]][["path"]]
+  # replace right-most name of library,
+  # should work across platforms
+  lib_path <- sub("(.*)RSQLite[.]", "\\1regexp.", lib_path)
```
Comment (krlmlr) [inline R/regularExpressions.R]:
```suggestion
  lib_path <- sub("(.*)RSQLite([.][a-zA-Z]+)$", "\\1regexp\\2", lib_path)
```

**Comment by rfhb** [comment]
> 
> 
> Thanks! Would it be too bad to statically link the extension? I'd assume this doesn't contribute a lot to the size of the shared library.

Many thanks - yes, this is a small library (less than 30 kB). Indeed, initially I compiled and statically linked the `regexp.c` extension, but loading the combined library threw an error at runtime because of a duplicate function definition for the entry point (if I remember correctly). I patched the source code of this extension and of the mathematical functions (`src/vendor/sqlite3/extension-functions.c`), trying to follow the [SQLite guidance on statically linking](https://www.sqlite.org/loadext.html#statically_linking_a_run_time_loadable_extension), but was not successful since I don't know enough about this subject. 

I thought it may be more acceptable to build a separate library than to modify the source (which may change with SQLite extending the amalgamation of sources). 

Could you help or advise?

File: R/regularExpressions.R (hunk 1 above)
Comment (rfhb) [reply R/regularExpressions.R]: Yes, thank you, now implemented in c6471f4.

File: R/regularExpressions.R (hunk 2 above)
Comment (rfhb) [reply R/regularExpressions.R]: Thanks, with changes (non-greedy second group) now implemented in c6471f4.

File: R/regularExpressions.R [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  lib_path <- getLoadedDLLs()[["RSQLite"]][["path"]]
+  # replace right-most name of library object (RSQLite) with
+  # name of the regexp object, should work across platforms
+  lib_path <- sub("(.*)RSQLite([.][a-zA-Z]+?)$", "\\1regexp\\2", lib_path)
+
+  # repeat loading would throw error
+  if (is.null(getLoadedDLLs()[["regexp"]][["path"]])) {
+
+    # try to load
+    res <- try(
```
Comment (krlmlr) [inline R/regularExpressions.R]: I don't mind throwing an error if this fails, just like `initExtension()` does. The code path could be simplified to "if not loaded then load; return true".

File: tests/testthat/test-regularExpressions.R [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  # note true only from first
+  # invocation of dbConnect()
+  expect_true(
+    initRegExp(
+      db = con))
+
+  dbWriteTable(
+    conn = con,
+    name = "iris",
+    value = iris)
```
Comment (krlmlr) [inline tests/testthat/test-regularExpressions.R]: Maybe we can use a custom dataset with a few strings and regexp-s that match some?

File: R/regularExpressions.R (hunk 3 above)
Comment (rfhb) [reply R/regularExpressions.R]: I see. On the other hand, why not providing a meaningful return value? Users could easier check if `REGEXP` is available or not. It's best in your hands as package maintainer.

File: tests/testthat/test-regularExpressions.R (hunk 4 above)
Comment (rfhb) [reply tests/testthat/test-regularExpressions.R]: Sure, what about the extended string examples added now with commit 22f7fbd?

**Comment by krlmlr** [comment]
Thanks!

**Comment by krlmlr** [comment]
There's a problem on Linux, `getLoadedDLLs()` doesn't seem to contain the regexp DLL even when loaded. From experiments with commenting parts of `sqlite3_regexp_init()` and from reading `sqlite3.c` it seems that reloading the extension is harmless when invoked from the C API [`sqlite3_load_extension()`](https://www.sqlite.org/c3ref/load_extension.html), we should try that (also for the default extensions).

**Comment by krlmlr** [comment]
Done. I'm also taking care of moving the extension to the main RSQLite shared library.

**Comment by rfhb** [comment]
That's great, thank you for finding a much better way to compile, link and load, also quite instructive for me. I think users will like the extension as well, thanks a lot for this!

**Pull request merged**